Add contextual SCP to SSHKit module

brienw opened the pull request:
We ran into the need for these context aware SCP operations, so I went ahead and implemented them. I tried to follow the spirit of what you had in the comments.

I've also moved my previous file upload fix to the `write` function instead of in the `next` function.

---

* [X] Documented the change if necessary
* [X] Tested this PRs change (with unit and/or functional tests)
* [X] Added this PRs change to CHANGELOG.md (in the `#master` section) if  necessary.

pmeinhardt commented:
Hey @brienw, thanks a lot for taking a stab at this and esp. for taking a look at what we drafted. We're at ElixirConf.EU this week so I'm really happy to not have to worry about the "daily business" and instead spend some time on SSHKit.

We'll be adding functional tests – our primary objective for the following evenings.

Then we can also add tests for this PR before merging, yay 😃 

As for your last message: I won't be at OSConf, but I would definitely be up for a video call with you guys. I'm really curious about the tool you're building and would love to hear where you're going with it, design goals etc. Is it open source/publicly available yet?

brienw commented:
@pmeinhardt Yes! I'm happy to hear about you guys having time to work on SSHKit this week and tackle functional tests. That will be awesome! I'd love to add tests to this once that's ready.

As for a video call, lets coordinate when you are back from the conf. We're building a tool called `bootleg` a pure elixir alternative/evolution of `edeliver` to help with build and deploy of elixir applications. It will be open source, but the repo is currently private. We just want to get the basic functionality it in a relatively clean/stable state before we open up the repo publicly as it's a moving target right now. Soon! When we talk we can grant repo access if we don't have it public yet. Would love to get more eyes/ideas on it.

Enjoy ElixirConf.EU, we have to wait many more months before we head up to ElixirConf here in the states

tessi commented:
I put up some functional tests. They are now merged into master. Find them at [test/sshkit_functional_test.exs](https://github.com/bitcrowd/sshkit.ex/blob/master/test/sshkit_functional_test.exs) as a reference.

Regarding `bootleg`: We had something similar in mind (which is why we started with sshkit as a first step). So great that you are going into that direction. We can't wait to learn more about `bootleg`  🚀

brienw commented:
I rebased this on top of #59 since it made sense to have the base SCP functional tests in place before I added all these context related tests. If #59 is merged, I'll double check this branch to make sure I don't need to rebase it again

pmeinhardt commented on lib/sshkit.ex:
```diff
@@ -262,17 +263,97 @@ defmodule SSHKit do
+  @doc ~S"""
+  Upload a file or files to the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file upload was successful.
```
I think it would be worth pointing out, that the list may also contain `{:error, reason}`.

How about also adding a reference to `SSHKit.SCP.upload/4` which is where each entry in the returned list comes from?

pmeinhardt commented on lib/sshkit.ex:
```diff
@@ -262,17 +263,97 @@ defmodule SSHKit do
+  @doc ~S"""
+  Download a file or files from the given context.
+  Returns a list of `:ok`.
+  There is one `:ok` per connected host where file download was successful.
```
See comment on line 269:

- pointing out that `{:error, reason}` may appear in the list if there's an error
- possibly add a reference to `SSHKit.SCP.download/4`

pmeinhardt commented on lib/sshkit.ex:
```diff
@@ -262,17 +263,97 @@ defmodule SSHKit do
+      false -> case context.path do
+        nil -> path
+        _ -> Path.join(context.path, path)
+      end
```
How about

```elixir
nil -> case context.path do
  nil -> path
  base -> Path.join(base, path)
end
```

because `base` will be bound to the value of `context.path`?

Maybe this is even something Credo should check? (Potentially open an issue there to detect when you're accessing the original value where you could be using the match?)

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+      assert [:ok] = SSHKit.download(context, remote, as: local)
+      assert File.read(shared) == File.read(local)
+      File.rm!(local)
+    end
```
Will wait to get your input on #59 before looking into the tests in more detail.

A few cases we may want to cover as well (could split this up into multiple PRs to move forward and have the contextual SCP in master sooner):

- [ ] test `:recursive` upload
- [ ] test `:recursive` download
- [ ] test `:preserve` upload
- [ ] test `:preserve` download

pmeinhardt replied:
…and this is really testing `SSHKit.download/3` not `SCP.download`, so it should probably move to `test/sshkit_functional_test.exs`.

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+      assert [:ok, :ok] = SSHKit.upload(context, local, as: remote)
+      assert File.read(local) == File.read(shared)
+      File.rm!(shared)
+    end
```
This is really testing `SSHKit.upload/3` so it should probably go into `test/sshkit_functional_test.exs`.

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+
+      assert File.read("#{@local_workspace}/#{local}") == File.read(shared)
+      File.rm!(shared)
+    end
```
Same here I think

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+      assert [:ok] = SSHKit.upload(context, local, as: remote)
+      assert File.read(local) == File.read(shared)
+      File.rm!(shared)
+    end
```
…and here 😄

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+      assert [:ok, :ok] = SSHKit.download(context, remote, as: local)
+      assert File.read(shared) == File.read(local)
+      File.rm!(local)
+    end
```
This is also testing `SSHKit.download/3`, so should go into `test/sshkit_functional_test.exs`, right?

pmeinhardt commented on test/sshkit/scp_functional_test.exs:
```diff
@@ -9,29 +9,141 @@ defmodule SSHKit.SCPFunctionalTest do
+
+      assert File.read(shared) == File.read("#{@local_workspace}/#{path}")
+      File.rm!("#{@local_workspace}/#{path}")
+    end
```
Same here

klappradla commented on lib/sshkit.ex:
```diff
@@ -262,17 +263,96 @@ defmodule SSHKit do
+    end
+
+    Enum.map(context.hosts, run)
+  end
```
Hey @brienw, @tessi and me had a look at `upload` and came up with a bunch of ideas for refactoring it for better understanding and readability. We'd suggest to split off a `build_remote_path` function which can also be reused within `download`.

Here's what this could look like:

```elixir
def upload(context, path, options \\ []) do
  as_path = Keyword.get(options, :as, path)
  remote_path = build_remote_path(context, as_path)

  run = fn host ->
    {:ok, conn} = SSH.connect(host.name, host.options)
    res = SCP.upload(conn, path, remote_path, options)
    SSH.close(conn)
    res
  end

  Enum.map(context.hosts, run)
end

defp build_remote_path(context, path) do
 if Path.type(path) == :absolute
   path
 else
   Path.join(context.path || “”, path)
 end
end
```

`upload` now also closes the connection per host - something that for now may seem a bit tedious, but will probably be addressed when it comes to #41 ✌️ 

We'd also change the documentation to this:

    @doc ~S"""
    Upload one or many files to the hosts of the given context.

    Returns a list of upload-result-tuples -- one for each host.
    Each result tuple is either `{:ok, output, status}` or {:error, “reason”}.

    Possible options are `as: "new_name"` (specify the name of the uploaded file/directory)
    and all options specified in `SSHKit.SCP.Upload.transfer/4`.

    ## Examples

    Upload all files and folders in the current directory
    to the remote directory “/workspace”:
       
    ```
    [{:ok, _, _}] =
      "my.remote-host.tld"
      |> SSHKit.context
      |> SSHKit.path("/workspace")
      |> SSHKit.upload(".", recursive: true)
    ```

    Uploads a file with a different name to the target host:

    ```
    [{:ok, _, _}] =
      "my.remote-host.tld"
      |> SSHKit.context
      |> SSHKit.upload("some_file.txt", as: "other_file.txt" )
    ```

    """

It especially changes the part on what is returned by `upload`. We *think* 🙈  that this would be a list of either `{:ok, output, return_value}` or `{:error, "reason"}` tuples rather then only a list of `:ok` atoms.
Is this assumption correct?

klappradla replied:
Sure that's outta the scope of this PR. I'm pretty new to the project, but one should probably put some thought into how various functions from the DSL will be combined and based on that decide what would make sense as return values.

I could for instance see the following approaches:

**a) Return `{:ok, conn, _x}`**
This would allow chaining commands like this:
```elixir
with {:ok, conn}      <- SSHKit.SSH.connect("foo.io"),
     {:ok, conn, :ok} <- SSHKit.SCP.upload(conn, "local", "remote"),
     {:ok, conn, :ok} <- SSHKit.SCP.upload(conn, "other_local", "remote"),
     {:ok, conn, res} <- SSHKit.SSH.run(conn, "command"),
     {:ok, conn, :ok} <- SSHKit.SSH.close(conn)
do
     {:ok, res}
else
     err -> err
end
```

**b) Return a connection with certain fields set**
This would be similar to how [Plug.Conn](https://hexdocs.pm/plug/Plug.Conn.html#content) works afaik. Building workflows could then look like this:
```elixir
"foo.io"
|> SSHKit.SSH.connect()
|> SSHKit.SCP.upload("local", "remote")
|> SSHKit.SSH.run("command")
|> SSHKit.SSH.close()
```

I'm personally more in favor of building workflows with "with", since it allows a better / more fine granular control over how errors are handled, but that comes down to personal preference and the general vision of the DSL I guess.

klappradla commented on lib/sshkit.ex:
````diff
@@ -262,17 +263,96 @@ defmodule SSHKit do
+  ```
+
+  """
+  def download(context, path, options \\ []) do
````
In order to align this with the changes we proposed for `upload`, this could look somewhat along the lines of this:
```elixir
def download(context, path, options \\ []) do
  remote = build_remote_path(context, path)
  local = Keyword.get(options, :as, Path.basename(path))

  run = fn host ->
    {:ok, conn} = SSH.connect(host.name, host.options)
    res = SCP.download(conn, remote, local, options)
    SSH.close(conn)
    res
  end

  Enum.map(context.hosts, run)
en
```
It would then also make use of the previously defined `build_remote_path` function we proposed in the changes for `upload` 🎉 

Same as `upload`, it also closes the connection per host - something that for now may seem a bit tedious, but will probably be addressed when it comes to #41 ✌️ 

And for the docs, same as for `upload`, depending on whether or not we have correct assumptions on the return values of `SCP.download` 😉

brienw commented:
Awesome feedback, thanks! I've been letting this branch collect way too many cobwebs while working on the other SCP tests branch. I'll try to raise this branch up from the dead by the end of the week and get your refactor suggestions included.

brienw commented:
Not sure what's up with that failing test but i'll look into it to see if i can figure it out. I can not get it to fail locally, so I wonder if it's a gremlin

```sh
mix test --seed 470257                             1 ↵  3549  18:33:13
..................................................................

Finished in 48.2 seconds
66 tests, 0 failures

Randomized with seed 470257
```

brienw commented:
Conflicts resolved and tests are passing now.

klappradla commented on lib/sshkit.ex:
```diff
@@ -319,19 +320,91 @@ defmodule SSHKit do
+  end
+
+  defp build_remote_path(context, path) do
+    if Path.type(path) == :absolute do
```
Just personal preference, but functions only consisting of one `if` - `else` structure always appear a bit suspicious to me 😉 

Can't the problem be solved via [`Path.absname/2`](https://hexdocs.pm/elixir/Path.html#absname/2) alone?
I just tried it out in iex and it looks like this:
```
Interactive Elixir (1.4.5) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> abs_path = "/Users/max/file.txt"
"/Users/max/file.txt"
iex(2)> rel_path = "max/file.txt"
"max/file.txt"
iex(3)> abs_path = Path.absname(abs_path, "something_random")
"/Users/max/file.txt"
iex(4)> abs_path = Path.absname(rel_path, "")
"/max/file.txt"
iex(5)> abs_path = Path.absname(rel_path, "/Users")
"/Users/max/file.txt"
iex(6)>
```

So I think we could change the whole function to:
```elixir
defp build_remote_path(context, path) do
  Path.absname(path, context.path || "")
end
```

What do you think?

brienw replied:
with a minor tweak this works just dandy, thanks for the feedback!
`Path.absname(path, context.path || ".")`

klappradla commented on test/support/functional_assertion_helpers.ex:
```diff
@@ -0,0 +1,57 @@
+  def compare_command_output(context = %Context{}, local, remote) do
+    Enum.map(context.hosts,
+      fn(h) ->
+        {:ok, conn} = SSH.connect(h.name, h.options)
```
Could we also make use of the new "lambda" *(still don't know if that's semantically correct...)* syntax here?

Something like this:
```elixir
SSH.connect h.name, h.options, fn(conn) ->
  compare_command_output(conn, local, remote)
end
```
This would spare the extra `SSH.close`, but I'm not sure if it makes the helper less complex after all... What do you think?

brienw replied:
i like reducing lines of code, so 👍

brienw commented:
This is ready for another look, I took some time yesterday to update with the two suggestions from @klappradla. Thanks!

tessi commented:
🎉
